lib/msal-core/src/PopTokenGenerator.ts

```diff
@@ -0,0 +1,153 @@
+        const publicKeyDigest = getStringFromArrayBuffer(publicJwkBuffer);
+
+        const publicKeyHash = Utils.encode(publicKeyDigest).substr(0, PopTokenGenerator.HASH_LENGTH);
+
+        // Save key pair in cache
+        await this._cache.put(publicKeyHash, {
+            publicKey,
+            privateKey: unextractablePrivateKey,
+            requestMethod: request.requestMethod,
+            requestUri: request.requestUri
+        });
+
+        const popTokenString = JSON.stringify({
+            kid: publicKeyHash
+        })
+
+        return popTokenString;
+    }
+
+    /**
+     * Signs an access token with the given pop key.
+     * @param publicKeyHash Public key to sign payload with
+     * @param payload Payload to sign
+     */
+    async signToken(accessToken: string): Promise<string> {
+        const {
+            cnf: {
+                jwk: {
+                    kid: publicKeyHash
+                }
+            }
+        } = Utils.extractIdToken(accessToken);
+
+        const {
+            publicKey,
+            privateKey,
+            requestMethod,
+            requestUri
+        } = await this._cache.get(publicKeyHash);
+
+        const publicKeyJwk = await this._crypto.exportKey(publicKey, KeyFormat.jwk);
+        const publicJwkString = getJwkString(publicKeyJwk);
+
+        const header = {
+            alg: publicKeyJwk.alg,
+            type: KeyFormat.jwk,
+            jwk: JSON.parse(publicJwkString)
+        };
+
+        const encodedHeader = Utils.base64EncodeStringUrlSafe(utf8Encode(JSON.stringify(header)));
+
+        const urlComponents = Utils.GetUrlComponents(requestUri);
+        const uriPath = (urlComponents.AbsolutePath || "").replace(/\/+$/, "");
+        const uriHost = urlComponents.HostNameAndPort || "";
+        const uriQueryString = urlComponents.QueryString || "";
+
+        const pS256 = await this._crypto.digest(HashAlgorithm.sha256, uriPath);
+        const qS256 = await this._crypto.digest(HashAlgorithm.sha256, uriQueryString);
+
+        const payload = {
+            at: accessToken,
+            ts: Utils.now(),
+            m: requestMethod.toUpperCase(),
+            u: uriHost,
+            "p#S256": Utils.base64EncodeStringUrlSafe(getStringFromArrayBuffer(pS256)),
+            "q#S256": Utils.base64EncodeStringUrlSafe(getStringFromArrayBuffer(qS256))
+        }
+
+        const encodedPayload = Utils.base64EncodeStringUrlSafe(utf8Encode(JSON.stringify(payload)));
+
+        const tokenString = `${encodedHeader}.${encodedPayload}`;
+        const tokenBuffer = getArrayBufferFromString(tokenString);
+
+        const signatureBuffer = await this._crypto.sign(privateKey, tokenBuffer);
+        const encodedSignature = Utils.base64EncodeStringUrlSafe(getStringFromArrayBuffer(signatureBuffer));
+        const signedToken = `${tokenString}.${encodedSignature}`;
+
+        return signedToken;
```

Comment:
decent amount of awaits here, is there any errors that could come through the control flow to here?

Reply:
Certainly, and `UserAgentApplication` will have to catch any errors and fail gracefully.